advanced_topics.qmd

```diff
@@ -5,13 +5,114 @@
+
+In [Chapter @sec-datamanipulation], we mentioned that the arrow R package transforms dplyr syntax into Arrow Expressions which can be understood by Acero, and here we're going to dive into that a little deeper.
+
+Acero, arrow's query engine, has a large number of functions implemented for performing calculations on data.
+The complete list can be found in the [project documentation](https://arrow.apache.org/docs/cpp/compute.html#available-functions).
+
+Some of these functions have the same names are the R equivalent, for example, both Acero and R have a `min()` function which can be used to find the lowest value in an vector or array of values.
+
+Other functions have slightly different names; whereas R has `sd()` for calculating the standard deviation of a vector, Acero has `stddev()`.
+
+| Feature                        | R (`sd`)                                                | Acero (`stddev`)                                                                        |
+|-------------------|----------------------|--------------------------------|
+| **Function Name**              | `sd`                                                    | `stddev`                                                                                |
+| **Function Options**           | `na.rm = FALSE`                                         | `ddof = 0`, `skip_nulls = FALSE`, `min_count = 1`                                       |
+| **Purpose**                    | Computes standard deviation                             | Computes standard deviation                                                             |
+| **Handling of Missing Values** | `na.rm`: Boolean, removes `NA` values (default `FALSE`) | `skip_nulls`: Boolean, removes `NULL` values (default `FALSE`)                          |
+| **Degrees of Freedom**         | Not available                                           | `ddof`: Integer, denominator degrees of freedom (default `0`)                           |
+| **Minimum Count**              | Not available                                           | `min_count`: Integer, returns `NULL` if fewer non-null values are present (default `1`) |
+
+There are also some differences in the options for these functions; while R's `sd` function has arguments `x`, the value to be passed in and `na.rm`, whether to remove `NA` values or not, Acero's stddev also has `ddof` - denominator degrees of freedom, `skip_nulls`, the equivalent of `na_rm`, and `min_count` - if there are fewer values than this, return `NULL`.
+
+This also highlights the slight different in terminology here; whereas R has separate concepts of `NA` and `NULL`, Arrow considers them to be the same things.
+
+You don't usually have to worry about these differences yourself though, as the code in the arrow R package does a direct translation from one to the other.
+Let's take a closer look at an example.
+
+Say we wanted to only return data for states whose names start with the letter "w".
```

Comment:
Do you think we could come up with an example that isn't easily done with regular R functions? `utf8_swapcase` is a fun one that I don't think has an R/stringr version. `xor` is also interesting, if we had a non-contrived use case. There are other string functions that could be done with regular expressions but might be convenient not to, like `utf8_is_alnum`, `utf8_is_decimal`, `utf8_is_digit`, etc.

FWIW those probably also don't involve having to hunt down compute options, so that would make them easier to show. We could also show one that uses compute options, but unless we've had to add bindings for the function, we probably haven't wired up the options in `make_compute_options`, so it probably wouldn't just work anyway.